src/Controller/ArticlesController.php

```diff
@@ -696,6 +700,76 @@ public function risAction(Request $request, string $id) : Response
         return new Response(preg_replace('~\R~u', "\r\n", $this->get('templating')->render('::article.ris.twig', $arguments)), Response::HTTP_OK, ['Content-Type' => 'application/x-research-info-systems']);
     }
 
+    public function xmlAction(Request $request, string $id, int $version = null) : Response
+    {
+        $arguments = $this->defaultArticleArguments($request, $id, $version);
+
+        $xml = $arguments['item']
+            ->then(Callback::method('getXML'))
+            ->wait();
+
+        $xForwardedFor = array_filter(array_map('trim', explode(',', $request->isFromTrustedProxy() ? $request->headers->get('X-Forwarded-For') : '')));
```

Comment:
I agree a streaming helper will help consolidate the code here in a single place (which has been flagged for duplication).

It seems the dependencies it should take are `$request`, `$xml` (which generically we could call `$url`), returning the `StreamedResponse`.

Later on we are going to try to outsource this to Fastly, and that will be made simple by having a single helper that contains it.

Reply:
Unrelated: I always forget why we stream files from here. Maybe related to analytics or to hiding the URL we are loading.

Reply:
Ah right, it was about setting custom headers on the response (`Content-Disposition`?).